### PR TITLE
[ARM] `az deployment`: Fix parse errors with multiline strings

### PR DESCRIPTION
**Pull request by wiktorn**
**Description**  
For more complex multiline strings (such as policies for Azure API Management), `az deploy` will fail with message: ` Failed to parse ... , please check whether it is a valid JSON format`, because multiline string contains block quotes within the string.

This PR adds test case that's failing with current az and naive proposal to improve detection of such situation. I couldn't find a regexp that is solving all cases, that's why I just slightly changed existing one and added check with replacement. The regex is already quite complex and I think it's already on the bridge of being understandable.

**Testing Guide**  
Test cases are provided within PR.

**History Notes**  
[resource] Fix parsing JSON files with multi-line strings  

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
resource

**Comment by zhoxing-ms** [comment]
Related issue: #12909

**Comment by zhoxing-ms** [comment]
@wiktorn Hi, thank you for your contribution.

However, there are still some problems with the support of service for multi-line, the verification of CLI pass multi-line still can't solve the problem.

1. If it is multi-line of pure describing text, service will break the original format with newline character. For example:
- In ARM template:
```
"meta": {
    "type": "string",
    "metadata": {
        "description": "meta of the network security group."
    },
    "defaultValue": "ARM
    template
    test
    newline"
}
```
- After service processing:
```
"meta": {
    "type": "String",
    "value": "ARM\r\n      template\r\n      test\r\n      newline"
}
```

2. If it is multi-line of parameter values, service may break the resolution of those values.
For example:
```
"defaultValue": "
    [resourceGroup().location]
"
```
It will cause an error in the service:
```
{"error":{"code":"InvalidResourceLocation","message":"The specified location '\r\n        [resourceGroup().location]\r\n      ' is invalid. A location must consist of characters, whitespace, digit, or following symbols '(,)'."}}
```

Please refer to this issue #12909 for the discussion of the service problem.
I suggest that this PR should not merge at this time. After the service solves this problem, this PR can be merged to pass the verification of multi-line in CLI side.

**Comment by wiktorn** [comment]
I gave this second thought and I thing that maybe the best approach to this problem is to use custom json decoder (lets call it azjson) that would respect the rules for Resource templates (comments allowed, multiline strings allowed). Is there any chance to use the same algorithm on both ends?

If such tool would be written, then following template:
```
"meta": {
    "type": "string",
    "metadata": {
        "description": "meta of the network security group."
    },
    "defaultValue": "ARM
    template
    test
    newline"
}`
```
Would indeed have in-memory representation as:
```
"meta": {
    "type": "String",
    "value": "ARM\r\n      template\r\n      test\r\n      newline"
}
```
(modulo platform discrepancies around CRLF/CR). 

Also - taking a look into the json module, maybe far easier approach than this regex magic would be just to pass `strict=False` argument to json parser in `shell_safe_json_parse`? The difference would be, that other control characters would be allowed within template.

**Comment by zhoxing-ms** [comment]
@wiktorn Hi, may I ask what should we do if the multi-line is in the value of parameters?
Such as: 
```
"defaultValue": "
    [resourceGroup().location]
"
```
If it is translated to 
```
defaultValue": "\r\n        [resourceGroup().location]\r\n      "
``` 
, the parsing of this value should be destroyed.

And another factor is that in our design, CLI can't modify the user's template, so it needs to send the template directly to the Service.

**Comment by wiktorn** [comment]
> @wiktorn Hi, may I ask what should we do if the multi-line is in the value of parameters?
> Such as:
> ```
> "defaultValue": "
>     [resourceGroup().location]
> "
> ```

With such templates - do noting. Pass it to Service, and if it will be rejected there - it's perfectly fine as the message is clear (location should not contain whitespace characters) and user can easily fix.

As I understand, you would like multiline strings behave something like that (example in Python):
```python
defaultValue = "" +
    "[resourceGroup().location]" +
    ""
```

But even multiline strings in Python do not behave that way and maintain all the whitespace even though it has semantic meaning in language and for example - all it is possible to remove whitespace that was used to indent nearby code. But this was only implemented (as post-processing) to [docstrings](https://www.python.org/dev/peps/pep-0257/). So I think that Pythons strategy with multiline strings is perfectly fine here too and it's still in line with the documentation. AFAIR there is nothing there, that some magic with whitespace will happen.

Or maybe I'm just totally not understanding the problem that you raise with this. My assumptions are:
* such template fails on Service, because location contains invalid characters
* there is no implementation of ARM Template functions, so cli doesn't parse the value

**Comment by zhoxing-ms** [comment]
> such template fails on Service, because location contains invalid characters

Yes, the value cannot be parsed properly because it contains invalid characters, such as `\r\n      `.

**Comment by wiktorn** [comment]
As I see no follow-up on #12909, can we proceed with this one? Would you like to stay with this approach or you prefer to go with `json.load(..., strict=False)` ?

**Comment by zhoxing-ms** [comment]
Let me try to contact the service team again. I have just summarized the questions into an email to ask them

**Comment by wiktorn** [comment]
@zhoxing-ms any updates on this?

**Comment by zhoxing-ms** [comment]
@wiktorn Hi, as we discussed with the Service team, the current behavior of the service is in line with expectations, so we only need to merge this PR to solve the CLI verification problem (If you are interested in the details of our communication, I can forward the email to you~)

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 1]
```diff
@@ -287,6 +287,17 @@ def _remove_comments_from_json(template, preserve_order=True, file_path=None):
         raise CLIError("Failed to parse the JSON data, please check whether it is a valid JSON format")
 
 
+def _remove_multiline_from_json(template):
+    result = template
+    # find all text within bock quotes (single-line and multiline)
+    for match in re.finditer(r'((?<!\\)")((?:.?(?!(?<!\\)"))*.?)??"', template, re.DOTALL):
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:293]: In order to increase the readability of the code, could you please add some comments explaining the regular expression in detail? For example, add some matching samples to it?

**Comment by wiktorn** [comment]
I've created the branch with the other approach (non-strict JSON parsing) which doesn't need any magic regexes:
https://github.com/Azure/azure-cli/compare/dev...wiktorn:multiline_string_nonstrict_parse?expand=1

If you'd still want to continue with present approach, I've added comments with description how this regex works.

**Comment by wiktorn** [comment]
Shall I rebase it on dev? There are conflicts in recordings, as I've recorded them using WSL, also - not sure what is the policy to provide them.

OTOH, as far as I see, there are some other tests failing currently on dev.

**Comment by zhoxing-ms** [comment]
> I've created the branch with the other approach (non-strict JSON parsing) which doesn't need any magic regexes:
> https://github.com/Azure/azure-cli/compare/dev...wiktorn:multiline_string_nonstrict_parse?expand=1
> 
> If you'd still want to continue with present approach, I've added comments with description how this regex works.

It's a nice solution ! I think it's more readable and easier to maintain~
Could you please use it to update this PR?

**Comment by zhoxing-ms** [comment]
> Shall I rebase it on dev? There are conflicts in recordings, as I've recorded them using WSL, also - not sure what is the policy to provide them.
> 
> OTOH, as far as I see, there are some other tests failing currently on dev.

Yes, you need to fetch the latest code from the remote dev branch, then rebase locally before committing this PR

**Comment by wiktorn** [comment]
Code rebased, changed approach to non-strict json parsing.

I haven't updated the test recordings, not sure what to do about that.

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 2]
```diff
@@ -275,10 +275,8 @@ def _remove_comments_from_json(template, preserve_order=True, file_path=None):
     # It will affect the subsequent multi-line processing logic, so deal with this situation in advance here.
     template = re.sub(r'(^[\t ]*//[\s\S]*?\n)|(^[\t ]*/\*{1,2}[\s\S]*?\*/)', '', template, flags=re.M)
     minified = jsmin(template)
-    # Get rid of multi-line strings. Note, we are not sending it on the wire rather just extract parameters to prompt for values
-    result = re.sub(r'"[^"]*?\n[^"]*?(?<!\\)"', '"#Azure Cli#"', minified, re.DOTALL)
     try:
-        return shell_safe_json_parse(result, preserve_order)
+        return shell_safe_json_parse(minified, preserve_order, strict=False) # use strict=False to allow multiline strings
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:279]: Please solve the code style issues from Flake8 verification
> custom.py:279:77: E261 at least two spaces before inline comment

```suggestion
        return shell_safe_json_parse(minified, preserve_order, strict=False)  # use strict=False to allow multiline strings
```

**Comment by zhoxing-ms** [comment]
> Code rebased, changed approach to non-strict json parsing.
> 
> I haven't updated the test recordings, not sure what to do about that.

@wiktorn In fact, you need to rerun the test `azdev test test_resource_group_deployment --live` in live mode and then upload the updated yaml file.

**Review by zhoxing-ms — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]

**Pull request merged**